### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by btkostner**
Time for another cleanup branch!

This branch was tested, with Jenkins and Aptly. All is working as expected.

### Changes:
- Move `app.js` to `lib/` directory
  - Allows more modular use of components. Flightcheck no longer initializes a database because it never uses one.
  - Easier testing of modules
- Added an `atc` library that wraps socket.io for internal communication between processes. Includes message queue and multiple connection handling (picks a random connection currently)
- Added `mistake` class. Extends a normal error, but it includes an http code. Results in better error handling like 'Houston was unable to communicate with GitHub' vs 'Houston has encountered an error'
- Added a `render` library that wraps nunjucks. Separates title from body and removes weird extra new line stuff for GitHub issues.
- Rename `core` to `houston` and `appHooks` to `flightcheck`. Purely a cosmetic change.
- Centered everything on the error page. Looks slightly nicer, but it still needs work before release.
- Stricter linting rules. Includes always use parentheses for arrow functions, chain function length, no console logs, and a few others.
- Removed some useless dependencies. Replaced some plugins with inhouse code.
- More global use of bluebird for promise support. This includes mongoose and superagent
- Added unit tests. This ones a biggy.
  - All `lib/` files have associated tests to make sure they work the way they should be
  - Project linting is a test now
  - Tests all flightcheck templates to make sure they don't have syntax errors
  - Tests all houston policies to make sure that security functions work
  - Some other tests for houston (aptly and atc service)

### Fixes
- #148 GitHub organization permission checking
- #149 Multiple flightcheck instances testing a cycle, and reporting back, causing double issues to be submitted. (Still want to lock down api more, so after information is sent back, no other information is accepted.)

### CODE REVIEW

File: lib/atc.js [hunk 1]
```diff
@@ -0,0 +1,184 @@
+/**
```
Comment (fabianthoma) [inline lib/atc.js]: Why the rename to atc, does it have any specific meaning or is it just to fit the theme of flight related names?

File: entry.js [hunk 2]
```diff
@@ -6,9 +6,12 @@
 require('babel-register')
 require('babel-polyfill')
 
+// The code forced me to overwrite native functions. I'm sorry
+global.Promise = require('bluebird')
```
Comment (fabianthoma) [inline entry.js]: It's ok, we forgive you. Because bluebird is the best promise implentation I've seen so far :+1:

File: .gitignore [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # Houston files
 *.log
-config.js
+/config.js
```
Comment (fabianthoma) [inline .gitignore]: Why did you change this path?

File: lib/atc.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ atc.connected = false
 atc.holdMessages = []
 atc.id = crypto.randomBytes(16).toString('hex')
 atc.sentMessages = []
-atc.type = 'server'
```
Comment (fabianthoma) [inline lib/atc.js]: Why default to server, should this not be a parameter from whoever is calling this library?

File: .gitignore (hunk 3 above)
Comment (btkostner) [inline .gitignore]: There is another config file for tests. I only want the root config to be ignored.

File: entry.js [hunk 5]
```diff
@@ -6,24 +6,36 @@
 require('babel-register')
 require('babel-polyfill')
 
-var Path = require('path')
-
-var currentIndex = process.argv.indexOf(__filename)
+var path = require('path')
+
+// The code forced me to overwrite native functions. I'm sorry
+global.Promise = require('bluebird')
```
Comment (fabianthoma) [inline entry.js]: it's ok, bluebird is awesome anyway

File: flightcheck/appHook.js [hunk 6]
```diff
@@ -50,13 +55,12 @@ class AppHook {
     this.information = Object.assign(this.information, obj)
   }
 
-  // TODO: flatten data object
-  // TODO: base64 decode all files automaticly?
   file (path) {
-    return Request
+    return request
     .get(`https://api.github.com/repos/${this.data.project.github.fullName}/contents/${path}?ref=${this.data.tag}`)
```
Comment (fabianthoma) [inline flightcheck/appHook.js]: are we still not checking the code out? Or is this in one of your newer branches?

Comment (btkostner) [inline flightcheck/appHook.js]: That will be later in the strongback branch

File: flightcheck/appHook.js [hunk 7]
```diff
@@ -68,27 +72,19 @@ class AppHook {
...
-    let template = issue.render(`${this.path}/${this.mark}`, this)
+    const template = render(path.join(this.path, this.mark), this)
 
-    template = template.split('\n')
-    let title = template.find(string => string !== '')
-
-    template.splice(0, template.indexOf(title) + 1)
-
-    template = template.join('\n')
+    template.body += `\n<!-- Houston version ${config.houston.version} in ${config.env} environment -->`
```
Comment (fabianthoma) [inline flightcheck/appHook.js]: Is this version autogenerated or do we need to manually update it?
Cause tracing errors back to the commit level would be kind of nice.

Comment (btkostner) [inline flightcheck/appHook.js]: that is an assigned variable from the `package.json` file. So what ever version is in that will be the version outputted. I could try to check for a `.git` folder and try to read the latest commit sha and append that to the version number so it's like `1.0.0-as93jfs` if you think that would be helpful.

File: houston/controller/hook/index.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+const route = new Router({
+  prefix: '/hook'
+})
+
+route.use(github.routes(), github.allowedMethods())
+route.use(jenkins.routes(), jenkins.allowedMethods())
+
+// Use event hook listeners as well
+require('./atc')
```
Comment (fabianthoma) [inline houston/controller/hook/index.js]: Why is this here, it's not used in this file. Is this to set up the routing for the websockets?

File: houston/controller/hook/jenkins.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+route.get('*', async (ctx, next) => {
+  if (!config.github.hook) {
+    throw new ctx.Mistake(503)
+  }
+
+  await next()
+})
+
+// TODO: redo this whole handler, it's messy and unoptimized, plus I think it secretly hates Linux
```
Comment (fabianthoma) [inline houston/controller/hook/jenkins.js]: What? :stuck_out_tongue_winking_eye:

File: houston/index.js [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  if (app.env === 'test') return
+
+  if (/4.*/.test(error.status)) {
+    log.verbose(`${ctx.method} ${ctx.status} ${ctx.url} |> ${error.message}`)
+  } else {
+    log.error(error)
+  }
+
+  try {
+    await next()
```
Comment (fabianthoma) [inline houston/index.js]: What exactly would be the next error handler, and why exactly are you waiting on it to return a promise?

File: flightcheck/appHook.js (hunk 7 above)
Comment (fabianthoma) [inline flightcheck/appHook.js]: Yeah I think we should note that down as a additional feature for logging, I'm filing an issue for that

File: houston/controller/hook/index.js (hunk 8 above)
Comment (btkostner) [inline houston/controller/hook/index.js]: It loads all the websocket routes for atc. Technically it's not a koa route, but it still does the same type of logic a controller does so it makes sense being here. It also doesn't export anything which is why it's just a `require()` instead of `import`.

**Comment by fabianthoma** [comment]
Damn this is such a fucking huge PR, I'm currently at 2/3 and I'm seriously exhausted just from looking at that code

File: houston/index.js (hunk 10 above)
Comment (btkostner) [inline houston/index.js]: ~~This is overwriting the built in error handler that koa has. The `next` is a possible hander for the error (I've never actually encountered one, I assume koa has some for built in errors), and if the error handler fails, then we have failed epicly. Keep in mind that the try also catches a non exist handler.~~

It's a catch all, 'oh god everything broke', guarantee of sending back information, 500 error page.

In order for something this bad to happen something like our view rendering would have to fail.

File: tests/index.js [hunk 11]
```diff
@@ -3,12 +3,28 @@
  * Entry for handling all Houston tests
  */
 
-require('babel-register')
```
Comment (fabianthoma) [inline tests/index.js]: whoo, no babel anymore?

**Comment by fabianthoma** [comment]
I'm going ahead and merging this one, I'm sure I missed quite a few things, but there is no point in anyone else going through that huge diff again.
